epyqlib/devicetreeview.py

```diff
@@ -231,6 +242,70 @@ def add_device(self, bus, device=None):
 
         return device
 
+    def check_software_hash(self, node):
+        pcan_bus = can.interface.Bus(bustype=node.interface, channel=node.channel)
+
+        # Recieve N messages on the PCAN bus.
+        # Discover the node ID from the arbitration ID.
+        # Store the found node ID's in a set.
+        node_id_set = set()
+        for i in range(100):
+            recv_id = pcan_bus.recv(0.1)
+            if recv_id is not None:
+                id = recv_id.arbitration_id & 0xFF
+                node_id_set.add(id)
+
+        # Set up the device.
+        factory_epc_path = pathlib.Path(epyqlib.tests.common.devices["factory"])
+        device = epyqlib.hildevice.Device(
+            definition_path=factory_epc_path,
+        )
+        device.load()
+        bus_proxy = epyqlib.busproxy.BusProxy(
+            bus=pcan_bus,
+            auto_disconnect=False,
+        )
+        device.set_bus(bus_proxy)
+
+        # Retrieve the software hash from the device.
+        # Display a dialog with the discovered device information.
+        hash = self.get_software_hash(device)
+        hash.addCallback(self.pcan_software_hash_pop_up_dialog, node, node_id_set)
+        hash.addCallback(lambda ignored: pcan_bus.shutdown())
```

Comment:
I think that if you put the decorators on this function (and switch it to an `async def`) then you could avoid any of the callback stuff.  Also could then use a `try`/`except` for the shutdown (or maybe the bus has a context manager?  I forget).

If you want to learn 'classic Twisted' then I think you might want `hash.addBoth(lambda ignores: pcan_bus.shutdown()`.  That'll run it either on success or error.  And I always forget but that's somehow a bit different than `.addCallback()` and `.addErrback()` separately...  I think...

Reply:
Moving the decorators was good advice. Simplifies the code.
I did play around with the `addBoth`, `addCallback`, `addErrback `when I was trying to get this to work. I spent a lot of time and had a lot of difficulty.

Reply:
`:]` The word 'callback' isn't associated with the word 'hell' for nothing.